pkl-core/src/main/java/org/pkl/core/stdlib/base/ListNodes.java

```diff
@@ -785,6 +788,79 @@ protected Object eval(VmList self, Object initial, VmFunction function) {
     }
   }
 
+  public abstract static class iterate extends ExternalMethod2Node {
+    @Child private ApplyVmFunction2Node applyLambdaNode = ApplyVmFunction2NodeGen.create();
+    @Child @LateInit private TypeNode typeNode;
+
+    private TypeNode getTypeNode() {
+      if (typeNode == null) {
+        CompilerDirectives.transferToInterpreterAndInvalidate();
+        var section = VmUtils.unavailableSourceSection();
+        var booleanTypeNode = new BooleanTypeNode(section);
+        var unknownTypeNode = new UnknownTypeNode(section);
+        typeNode = new PairTypeNode(section, booleanTypeNode, unknownTypeNode);
+      }
+      return typeNode;
+    }
+
+    @Specialization
+    protected Object eval(VirtualFrame frame, VmList self, Object initial, VmFunction function) {
+      var iter = self.iterator();
+      var typeNode = getTypeNode();
+      var result = initial;
+      var loop = 0;
+
+      while (iter.hasNext()) {
+        var elem = iter.next();
+        var pairUnchecked = applyLambdaNode.execute(function, result, elem);
+        var pair = (VmPair) typeNode.executeEagerly(frame, pairUnchecked);
+        result = pair.getSecond();
+        loop++;
+        if (!(Boolean) pair.getFirst()) break;
+      }
+
+      LoopNode.reportLoopCount(this, loop);
+      return result;
+    }
+  }
+
+  public abstract static class iterateIndexed extends ExternalMethod2Node {
+    @Child private ApplyVmFunction3Node applyLambdaNode = ApplyVmFunction3NodeGen.create();
+    @Child @LateInit private TypeNode typeNode;
+
+    private TypeNode getTypeNode() {
+      if (typeNode == null) {
+        CompilerDirectives.transferToInterpreterAndInvalidate();
+        var section = VmUtils.unavailableSourceSection();
+        var booleanTypeNode = new BooleanTypeNode(section);
+        var unknownTypeNode = new UnknownTypeNode(section);
+        typeNode = new PairTypeNode(section, booleanTypeNode, unknownTypeNode);
+      }
+      return typeNode;
+    }
+
+    @Specialization
+    protected Object eval(VirtualFrame frame, VmList self, Object initial, VmFunction function) {
+      var iter = self.iterator();
+      var typeNode = getTypeNode();
+      var result = initial;
+      long index = 0;
+      var loop = 0;
```

Comment:
What's the reason these are separate?

Is this style preferable?
```suggestion
      var index = 0L;
      var loop = 0;
```

Reply:
No reason really. I just copied this code from the `fold` implementation. But good point.